### DISCUSSION TITLE
Ginga program hooks (Ginga as a replacement for DS9)

### DISCUSSION THREAD

**Discussion by erincfigura** [ejeschke/ginga#1026] [General]
Aloha!
I'm new to Ginga, but I'm investigating it as a possible replacement for our use of SAOImage DS9 in a number of applications.
I have a couple questions that I've been unable to find the answers to, so please bear with me.

We've got a couple of applications currently that are interactive in the terminal, but use a ds9 display instance for display and interaction.  Both our application and the DS9 event loops run simultaneously, so that the image view can be modified either from DS9 or from our application.

Is this kind of behaviour possible with Ginga?  I've been experimenting both with the ReferenceViewer and with the CanvasView.  The ReferenceViewer capabilities are exactly what I'm looking for, but i have not been able to find any documentation on accessing the viewer functions from outside its event loop.

CanvasView, on the other hand, looks to be capable of this, but apparently I would need to build the viewer from the ground up.
I've been looking at the ReferenceViewer code, and have some ideas how one might brute-force this, but I'm wondering if there's something more elegant.

So, to summarise, my questions:

1. Does ReferenceViewer allow for hooks from outside its event loop?
2. Is there a built-in way to easily re-create the full ReferenceViewer capabilities using CanvasView and the other basic tools?

Mahalo nui, I appreciate any help anyone can provide.

**Comment by ejeschke** [comment]
@cfigura, there are a couple of answers to this, depending on whether you mean "from another process" (perhaps running in a terminal) or whether you can add code to run in the ginga process.  For the former, you probably want to have a look at the "RC" (Remote Control) plugin for Ginga. There is some documentation on its use in the [docstring in the module](https://github.com/ejeschke/ginga/blob/master/ginga/rv/plugins/RC.py). This plugin can also be a good guide for writing your own custom remote control protocol using some other IPC that you prefer.

For the latter scenario, if you are able to take your code (e.g. something running in the terminal) and run it from Python, then you might consider writing your own Ginga plugin. It is not that hard and we would be happy to provide guidance. There is a chapter on it in the readthedocs documentation, and of course, lots of examples in the ginga source code plugins directory.  If you are worried about blocking the event loop, you might be able to use something like Python's multiprocessing module to run your code as a subprocess.

**Reply by ejeschke** [reply]
Just thought I'd mention that we use something like the RC plugin to call into Ginga from our observation control system here at Subaru Telescope to do several different types of scripted observation commands using custom plugins. We pass data in and get results out. An example is to peek up on the slit of a spectrograph: we ask the user to draw a rectangle around an object and click on the slit (which might be in a different image or not). The difference is calculated and sent back and translated into a telescope movement command.

And...sorry, I didn't get to your question number 2.
It's fairly easy to create a cut down version of the Reference Viewer with a custom selection of plugins.  But you can consider the CanvasView as like a small but full featured viewer in a widget. But it is more for building your own viewers without having to reinvent the wheel.  There is a framework provided by the reference viewer that is assumed to be present by the plugins, that is not there with just the widget. The sample programs in the examples directory give more of a sense of what you can do starting with the widget.

**Comment by erincfigura** [comment]
Mahalo nui, thank you so much for your help!  It sounds as if what you're doing at Subaru is very much like what we're using DS9 for at Gemini, and what I'm interested in using Ginga for.  I've been playing around with the RC plugin with some success (I can load an image!) but the canvas routines are not available.  While I can grab a handle on the canvas for "Image":
`canvas = viewer.canvas('Image')
the canvas object has no non-underbar properties: trying something like the example "canvas.clear()", or any other canvas method (such as set_color_map or rotate, etc) result in a '#UNDEFINED' response.

The fact that I'm not seeing even the 'clear' method suggests to me that I'm doing something wrong here. What am I missing?

**Reply by ejeschke** [reply]
So if you follow an example like [this](https://github.com/ejeschke/ginga/blob/master/ginga/examples/jupyter-notebook/Remote_control_of_a_Ginga_reference_viewer.ipynb), you should be able to do something like:
```python
canvas = viewer.canvas('Messier')
import numpy as np
x = np.arange(100)
y = x * x
points = list(zip(x.tolist(), y.tolist()))
canvas.add('path', points, color='green')
```
at the end of the notebook and see a green line running up the side of your image (if you have an image loaded in channel 'Messier') If there is no image loaded you should still see a plot.  Does this work for you?

**Reply by ejeschke** [reply]
I think you may not see methods suggested by the notebook or python because what you actually have is a remote connection stub using XML-RPC under the hood (in the client) and you are not actually connected to something that can show Python introspection on a remote Python object in a separate process on a (possibly) separate host.  This is also why you may see '#UNDEFINED' for certain responses that return None.

**Reply by ejeschke** [reply]
@cfigura, are you at Gemini North? Perhaps you could visit and I could demo our custom plugin to show you some of the remote control we are doing with it.

**Reply by erincfigura** [reply]
@ejeschke , I am indeed at Gemini North.  I'd love that opportunity to see how your plugin works!

Thanks for the info above, that helps a lot.

**Comment by ejeschke** [comment]
@cfigura, any late morning or afternoon would be fine. I will attempt to contact you offline.
